managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -365,6 +367,48 @@ public void operationFailed(MetaStoreException e) {
         return updateCursorPropertiesResult;
     }
 
+    @Override
+    public CompletableFuture<Void> setCursorProperties(Map<String, String> cursorProperties) {
+        Map<String, String> newProperties =
+                cursorProperties == null ? new HashMap<>() : new HashMap<>(cursorProperties);
+        long stamp = cursorPropertiesUpdateLock.writeLock();
```

Comment:
After a closer look. This lock will block admin handler thread, and the unlock depends on metadata callback thread, it has potential deadlock risk (which depends on if `setCursorProperties` is called on metadata callback thread).

I not quite certain about the purpose of this lock. Can you describe the non-thread-safe case here?

Reply:
@Jason918 
> I not quite certain about the purpose of this lock. Can you describe the non-thread-safe case here?

The perform `updateSubscriptionProperties` by Admin-API and others using `CursorProperties` to have data race, for example, store the list of delayed message index bucket snapshot in the [PIP-195](https://github.com/apache/pulsar/issues/16763), so we need to add the lock to ensure data consistency.

Reply:
> This lock will block admin handler thread, and the unlock depends on metadata callback thread, it has potential deadlock risk (which depends on if setCursorProperties is called on metadata callback thread).

Use the `ledger.getExecutor().executeOrdered(ledger.getName(), () -> {})` to make execute them on the same thread to instead of use the lock? It would be better?

Reply:
@coderzc I think timeout is not the best option here.

You can take a look at `MetadataCacheImpl#readModifyUpdateOrCreate`. It implements atomic update on metadata cache and supports concurrent writes. The main idea is update metadata store with version check and retry the op if there is concurrent updates.

Reply:
@Jason918 Thanks for your suggestion.

I use `ledger.getExecutor().executeOrdered(ledger.getName(), safeRun(() -> {})` to serial execution operation and check if exist `BadVersionException` and retry the operation.